Forest life

SaveliyLeontiev opened the pull request:
Simulation of life in forest

SaveliyLeontiev committed:
Main project add

SaveliyLeontiev committed:
Method showStomach in class Animal was added

SaveliyLeontiev committed:
Now we see information about eaten objects

vdemina commented on LifeOfForest/Eatable.h:
```diff
@@ -0,0 +1,13 @@
+
+@protocol Eatable <NSObject>
+
+- (int)giveCalories;
```
Название у метода не совсем правильное. Все-таки не give, а get. Но в Ojective-C обычно принято, когда метод возвращает какое-либо значение, опускать get. Т.е. нужно было написать - (int)сalories, а это уже @property (readonly) int сalories (только getter). Но если использовать такой способ, то нужно будет сделать кое-какие изменения в классах, реализующих протокол Eatable, относительно сalories.

SaveliyLeontiev replied:
А про какие изменения идет речь? То есть нужно было всем классам добавит свойство калорий и переписать доступ к ним?

vdemina replied:
Да

vdemina commented on LifeOfForest/Grass.m:
```diff
@@ -0,0 +1,33 @@
+
+@implementation Grass
+
+- (instancetype)init
```
Почему я не могу инициализировать траву с каким-нибудь другим именем?

SaveliyLeontiev replied:
Просто, так как симуляция дня проходит автоматически, то и имена давались случайные. Но завтра смогу дописать инициализатор по имени.

vdemina commented on LifeOfForest/Herbivore.m:
```diff
@@ -0,0 +1,25 @@
+
+@implementation Herbivore
+
+- (instancetype)init
```
Прочитай про designated convenience and initializers initializers. Видимо в твоем случае в классе Animal designated инициализатором будет initWithName:andCalories:. Таким образом тебе нужно будет переопределить метод init в том классе и вызвать из него initWithName:andCalories:, либо послать сообщение об ошибке, что методом пользоваться нельзя. Далее, если у класса Herbivore нет своих собственных designated инициализаторов, то таковым будет initWithName:andCalories:, поэтому из метода init класса Herbivore нужно вызывать [self initWithName:str andCalories:50], а уже из реализации initWithName:andCalories: в Herbivore нужно вызвать этот метод у super.

vdemina commented on LifeOfForest/Animal.m:
```diff
@@ -0,0 +1,66 @@
+
+@interface Animal ()
+
+@property int calories;
```
Для целых чисел нужно использовать NSInteger и NSUInteger для беззнаковых целых вместо int/long.

vdemina commented on LifeOfForest/Animal.m:
```diff
@@ -0,0 +1,66 @@
+{
+    NSLog(@"In stomach of %@", [self description]);
+    NSLog(@"Begin");
+    if ([self.stomach count] != 0)
```
Обращай внимание на code style. Все переносы и пробелы должны быть однородны.

vdemina commented on LifeOfForest/Animal.h:
```diff
@@ -0,0 +1,20 @@
+@property (readonly) NSString *name;
+
+- (void)eat:(id<Eatable>)obj;
+- (instancetype)initWithName:(NSString*)name andCalories:(int)calories;
```
and можно опустить

vdemina commented on LifeOfForest/Animal.h:
```diff
@@ -0,0 +1,20 @@
+
+@property (readonly) NSString *name;
+
+- (void)eat:(id<Eatable>)obj;
```
Обычно публичное API содержит элементы в следующем порядке:
свойсва (property)
методы класса
инициализаторы
остальные методы

vdemina commented on LifeOfForest/Forest.m:
```diff
@@ -0,0 +1,133 @@
+}
+
+
+- (BOOL)can:(id)obj1 eat:(id)obj2
```
Не очень удачное название

vdemina commented on LifeOfForest/Animal.h:
```diff
@@ -0,0 +1,20 @@
+
+@interface Animal : NSObject <Eatable>
+
+@property (readonly) NSString *name;
```
Строковые типы лучше использовать с атрибутом copy, это поможет избежать проблем в будущем. Но нужно быть внимательным, когда переопределяешь сеттеры и геттеры, а также инициализируешь объект, не забыть сделать копию через метод copy.

SaveliyLeontiev committed:
Добавлен инициализатор по имени для класса Grass
Переименован метод giveCalories на calories
Переименован метод can:eat: на canFirstCreature:eatSecondCreature:
Свойство calories в классе Animal изменено на свойство numberOfCalories

SaveliyLeontiev committed:
Добавлены константы калорий для основных классов

SaveliyLeontiev committed:
Убрана тестовая строка в main.m

SaveliyLeontiev committed:
Общие правки
